docs/CppInteroperability/Roadmap.md

```diff
@@ -0,0 +1,141 @@
+
+It is a goal not to have hidden performance traps. C++ APIs should have similar performance in Swift as C++ and should be called directly when possible. Imported APIs, especially containers, should not be implicitly converted to Swift types (in contrast to the automatic Objective-C bridging for NSArray and NSString). 
+
+It is a goal to preserve the semantics (including performance semantics) that an API was written with by understanding the API’s semantics and mapping it to a Swift idiom that represents the same thing. 
+
+Swift programmers have used C++ APIs since Swift 1.0 through an Objective-C or C bridging layer. It is important that programmers are able to *incrementally* remove these (potentially huge) bridging layers and start using their C++ APIs directly. To make this transition incremental, C++ interoperability must not change the way Swift imports any existing C or Objective-C APIs, even when C++ interoperability is enabled. 
+
+It is *not* a goal to import every C++ API into Swift. APIs that are imported by Swift should feel Swifty and idiomatic. There must be an easy way to add annotations so that APIs become importable, especially when updating APIs directly is not an option (because the source code is not available). 
+
+## The approach
+
+The approach for importing C++ APIs into Swift is derived from the fact that both our goals, and C++ as a language, necessitate that not all APIs are imported, and to import an API the user must provide semantic information in some cases. Further, in order to realize our goals (especially the goals of safety and clear, well defined mappings), it is impossible to import any arbitrary C++ API. APIs that are imported by Swift must fall into one of our understood categories or be annotated with semantic information that can be used to classify them into a Swift idiom. 
+
+In this section, the roadmap will discuss several of the most common C++ API patterns and how they might be imported by the Swift compiler using the goals outlined above.
+
+### Definitions
+
+**Reference types** have reference semantics and object identity. A reference type is a pointer (or “reference”) to some object which means there is a layer of indirection. When a reference type is copied, the pointer’s value is copied rather than the object’s storage. This means reference types can be used to represent non-copyable types in C++.
+
+**Owned types** “own” some storage which can be copied and destroyed. An owned type must be copyable and destructible. The copy constructor must copy any storage that is owned by the type and the destructor must destroy that storage. Copies and destroys must balance out and these operations must not have side effects. Examples of owned types include `std::vector` and `std::string`. 
```

Comment:
Swift semantics require that copies can be made all over the place, imported types must fit with these semantics until Swift does not require that types are copied all over.

Reply:
Would a type like the following be a owned type: (It owns some storage but not all)
```
struct S { 
   int &unownedReference;
   int ownedIntegerFiled; 
}
```

Reply:
Guess it is a trivial type as per the below definition. So owned type is a super set of trivial types right or are we expecting an explicit copy constructor (directly or transitively)? It may be worth clarifying.

Reply:
I was surprised not to see “value semantics” in this definition, even in a qualified way. Owned types are often value types, right?

Reply:
It's wrong to say that Swift semantics require copying all over the place.  The rampant copying in Swift is almost entirely an implementation artifact, and this fact is easy to demonstrate. Nothing in the Swift documentation gives any assurance of observable differences in address between two objects.  The only copies actually required by the language are those needed to make sure that mutations to one var/inout binding don't affect the values of any others.

Reply:
I'm confused again: the previous paragraph reads like it's talking about categories of Swift types, and this one reads like it's categorizing C++ types.

Reply:
>I was surprised not to see “value semantics” in this definition, even in a qualified way. Owned types are often value types, right?

@beccadax just to clarify, the notion of owned types here is only types that has some storage (because they could potentially internal pointers to it through some method). It doesn't necessary has to have "value semantics" if you are thinking of types that behaves like conceptual values e.g. like String, COW types (assuming their type parameters have value semantics) and the like.  E.g. the following would be "owned", and so is a smart pointer because it allocates and stores a ref count and copies that storage in copy constructors.

More semantically, the property that is required is whether a type has "some storage it copies" and can expose inner pointers to it. (It can totally have any others pointer, point to any storage and may have any side-effects.)  

// This is an owned type as per the above definition.
struct S { 
   StringRef name;
   int &integerRefence
   vector<int> intVector;
   ...  
}